### ISSUE TITLE
different game "modes" ?

### ISSUE THREAD

**Issue by elidoran** [tobspr/shapez.io#1142]
Right now the game has a single (fun) mode.

What do you think about some optional modes, such as:

1. speedrunner challenges
2. resource constrained challenges (instead of infinite resources)
3. space constrained challenges (instead of infinite map)

For speedrunner challenges, it could do things like:

1. have a different series of levels with different shapes to make
2. a mode with all tools unlocked from the beginning
3. don't allow placing a certain building, like painters, and instead have some pre-placed painters near the hub which must be used for all painting. Then, all planning must consider how to achieve everything as fast as possible by keeping those painters busy painting as much of the time as possible. Make other similar modes for other buildings, like cutters/mixers/stackers.
4. a "multiple hubs" mode with different hubs wanting different shapes than the other hubs, maybe they each have 10 unique shapes they want different amounts of and "winning" is when you've provided all shapes required to all hubs. The rest is all about planning on what to make when and where.
5. speedruns could also be resource constrained or space constrained

Perhaps a "custom game" window could allow configuring all these various options and perhaps saving them as named configs to reuse later. Add an option to show the game config in a summary screen before starting the new game. Then recording it would show the config to viewers.

For example, it would be fun to play a space constrained map with a hub at each map corner and each hub wants a different series of shapes (levels) than the other hubs. The space constraint would also means it's resource constrained. So, perhaps we wouldn't be able to make full belts of whatever we want and would have to instead plan it out better. Then, each hub would have its own "rewards" for completing its levels and they could be something new like unlocking more map space, or, increasing the size of resources so we can extract more from them. Maybe one Hub's levels reward more map space, another rewards more dye resources, another rewards more shape resources, and the fourth rewards with speed upgrades...then it's up to the speedrunner to decide which order they want to make shapes to get rewards to make their strategy work.

**Comment by EmeraldBlock** [comment]
Duplicate of #981

**Comment by elidoran** [comment]
Cool, diff, ideas over there. I love the Zachtronics games like SpaceChem, Infinifactory and Opus Magnum.

**Comment by EmeraldBlock** [comment]
Oh I see, this is more of a "change gameplay mechanics" rather than "complete specific goals"

**Comment by elidoran** [comment]
The other suggestion is about doing things like the Zachtronics games which is, for a specific input, you must produce an output. And, you measure how you did with three metrics. Speed, they call "cycles". Space used, including where things move/swing into. And the number of parts needed. It's fun because you can often create something in a small space, but, it takes a lot of cycles, or, you can create something fast but it covers a huge amount of space. And, it's fun to figure out the least number of parts required to produce the result.

Shapez.io could do all that. Each of its level goals and upgrade goals are a required output and it could restrict the input. It could track similar metrics for a play through. It's a fun idea. It doesn't match up exactly because Shapez.io isn't about making a single thing and then you're done. It keeps changing. But, it could track the overall of all the stuff built so far at the completion of each hub level. Then players would try to use the fewest cycles, least space, and least parts to complete each level/upgrade along the way.


My idea doesn't involve any of those metrics. At a basic level, it's "do the current mode but with a different series of shapes required per level", and, add an option to start with all stuff unlocked. This is about the challenge of figuring out which things to build in which order and change/swap what they're making to get thru the levels the fastest you can. Speedrunning. It uses all the current things available in the Shapez.io game. So, it is about completing specific goals. The same goal, produce shapes for the hub level.

Doing something like a restricted map or restricted resources also isn't a new game mechanic. It's simply setting a cap on how much map it creates or how many resources it places on the map. It would change the player's decision making process to figure out how to make things the fastest. The goal is still the same: produce a shape.

Restricting placing a building isn't a mechanic change because the game already keeps buildings locked until they're unlocked. I'm suggesting modes which never unlock a specific building, and it places a set amount of that locked building onto the map so players have to do the best they can with that limited number of that building. That seems like a fun challenge. We'd have to ensure those limited buildings were doing as much work as possible constantly, and focus on upgrades for them.

Some changes to the game mechanics I am suggesting:

1. more than one hub
2. hubs have different requirements (from each other) and only accept the shapes they want.
3. new rewards. right now rewards for completing hub levels are unlocking buildings/tools. And, upgrades reward improved speeds. I'm suggesting a mode where the rewards for completing shapes for hub levels are adding more map space or more resources (dye, or base uncolored shapes) or improving speeds. This would require players to make more decisions for their strategy to make all the shapes as fast as possible.

**Comment by 1000i100** [comment]
About resource limit, what about resources with : 
- amount maximum
- current stock
- regeneration ratio (stock*1.01/s for example)
- regeneration rate (stock+1/s for example)
- destroyed if empty Yes/No (no more regeneration if stock 0 is reached)

This way you can go slow and resilient, or fast but not for long.
